### PR TITLE
8340786: Introduce Predicate classes with predicate iterators and visitors for simplified walking

### PR DESCRIPTION
**Pull request by chhagedorn**
This patch introduces new predicate classes which implement a new `Predicate` interface. These classes represent the different predicates in the C2 IR. They are used in combination with new predicate iterator and visitors classes to provide an easy way to walk and process predicates in the IR.

### Predicate Interfaces and Implementing Classes 
- `Predicate` interface is implemented by four predicate classes:
  - `ParsePredicate` (existing class)
  - `RuntimePredicate` (existing and updated class)
  - `TemplateAssertionPredicate` (new class)
  - `InitializedAssertionPredicate` (new class, renamed old `InitializedAssertionPredicate` class to `InitializedAssertionPredicateCreator`)

### Predicate Iterator with Visitor classes
There is a new `PredicateIterator` class which can be used to iterate through the predicates of a loop. For each predicate, a `PredicateVisitor` can be applied. The user can implement the `PredicateIterator` interface and override the default do-nothing implementations to the specific needs of the code. I've done this for a couple of places in the code by defining new visitors:
- `ParsePredicateUsefulMarker`: This visitor marks all Parse Predicates as useful. 
  - Replaces the old now retired `ParsePredicateIterator`.
- `DominatedPredicates`: This visitor checks the dominance relation to an `early` node when trying to figure out the latest legal placement for a node. The goal is to skip as many predicates as possible to avoid interference with Loop Predication and/or creating a Loop Limit Check Predicate. 
  - Replaces the old now retired `PredicateEntryIterator`.
- `Predicates::dump()`: Newly added dumping code for the predicates above a loop which uses a new `PredicatePrinter` visitor. This helps debugging issues with predicates.

#### To Be Replaced soon
There are a couple of places where we use similar code to walk predicates and apply some transformation/modifications to the IR. The goal is to replace these locations with the new visitors as well. This will incrementally be done with the next couple of PRs.

### More Information
More information about specific classes and changes can be found as code comments and PR comments.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340786](https://bugs.openjdk.org/browse/JDK-8340786): Introduce Predicate classes with predicate iterators and visitors for simplified walking (**Sub-task** - P4)

### Reviewers
 * @abdelhak-zaaim (no known openjdk.org user name / role) Review applies to [7fa77140](https://git.openjdk.org/jdk/pull/21161/files/7fa77140bf801bd0660d4441fbe177811c364ea5)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21161/head:pull/21161` \
`$ git checkout pull/21161`

Update a local copy of the PR: \
`$ git checkout pull/21161` \
`$ git pull https://git.openjdk.org/jdk.git pull/21161/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21161`

View PR using the GUI difftool: \
`$ git pr show -t 21161`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21161.diff">https://git.openjdk.org/jdk/pull/21161.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21161#issuecomment-2371710500)

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -1464,7 +1464,8 @@ IfTrueNode* PhaseIdealLoop::create_initialized_assertion_predicate(IfNode* templ
                                                                    Node* new_stride, Node* control) {
   assert(assertion_predicate_has_loop_opaque_node(template_assertion_predicate),
          "must find OpaqueLoop* nodes for Template Assertion Predicate");
-  InitializedAssertionPredicate initialized_assertion_predicate(template_assertion_predicate, new_init, new_stride, this);
+  InitializedAssertionPredicateCreator initialized_assertion_predicate(template_assertion_predicate, new_init,
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:1467]: Needed to rename `InitializedAssertionPredicate` -> `InitializedAssertionPredicateCreator` to distinguish from the newly added `InitializedAssertionPredicate` class:
- `InitializedAssertionPredicate`: Represent an already existing Initialized Assertion Predicate in the IR.
- `InitializedAssertionPredicateCreator`: Class to create the IR nodes to represent an Initialized Assertion Predicate (one could return an `InitailizedAssertionPredicate` instance after the creation but that is not used/needed at the moment.)

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -4319,13 +4319,21 @@ void PhaseIdealLoop::mark_loop_associated_parse_predicates_useful() {
...
+
+  void visit(const ParsePredicate& parse_predicate) override {
+    parse_predicate.head()->mark_useful();
+  }
+};
+
 void PhaseIdealLoop::mark_useful_parse_predicates_for_loop(IdealLoopTree* loop) {
   Node* entry = loop->_head->as_Loop()->skip_strip_mined()->in(LoopNode::EntryControl);
-  const Predicates predicates(entry);
-  ParsePredicateIterator iterator(predicates);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:4325]: `PredicateIterator` + ``ParsePredicateUsefulMarker` combo to replace this class.

File: src/hotspot/share/opto/loopnode.cpp [hunk 3]
```diff
@@ -6378,14 +6423,10 @@ void PhaseIdealLoop::build_loop_late_post_work(Node *n, bool pinned) {
     // Move the node above predicates as far up as possible so a
     // following pass of Loop Predication doesn't hoist a predicate
     // that depends on it above that node.
-    PredicateEntryIterator predicate_iterator(least);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:6381]: `PredicateIterator` + ``DominatedPredicates` combo to replace this class.

File: src/hotspot/share/opto/predicates.cpp [hunk 4]
```diff
@@ -166,17 +166,6 @@ void PredicateBlock::verify_block() {
 }
 #endif // ASSERT
 
-// Walk over all Regular Predicates of this block (if any) and return the first node not belonging to the block
-// anymore (i.e. entry to the first Regular Predicate in this block if any or `regular_predicate_proj` otherwise).
-Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deoptimization::DeoptReason deopt_reason) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.cpp:171]: Replaced by new `RegularPredicateBlock` class which does the skipping.

File: src/hotspot/share/opto/predicates.hpp [hunk 5]
```diff
@@ -416,23 +542,207 @@ class InitializedAssertionPredicate : public StackObj {
...
+
+  // Skip all predicates by just following the inputs. We do not call any user provided visitor.
+  Node* skip_all() const {
+    PredicateVisitor do_nothing; // No real visits, just do nothing.
+    return for_each(do_nothing);
+  }
+
+  // Walk over all predicates of this block (if any) and apply the given 'predicate_visitor' to each predicate.
+  // Returns the entry to the earliest predicate.
+  Node* for_each(PredicateVisitor& predicate_visitor) const {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.hpp:564]: The goal is eventually to only have this method which does the check for what kind of predicate we face in the graph and replace all other places where we do these kind of checks together with visitors.

File: src/hotspot/share/opto/predicates.hpp [hunk 6]
```diff
@@ -416,23 +542,207 @@ class InitializedAssertionPredicate : public StackObj {
...
+    if (_parse_predicate.is_valid()) {
+      predicate_visitor.visit(_parse_predicate);
+    }
+    return _regular_predicate_block_iterator.for_each(predicate_visitor);
+  }
+};
+
+// Class to walk over all predicates starting at a node, which usually is the loop entry node, and following the inputs.
+// At each predicate, a PredicateVisitor is applied which the user can implement freely.
+class PredicateIterator : public StackObj {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.hpp:615]: Calling structure `PredicateIterator` -> `PredicateBlockIterator` -> `RegularPredicateBlockIterator`.

File: src/hotspot/share/opto/predicates.cpp [hunk 7]
```diff
@@ -90,27 +82,31 @@ Deoptimization::DeoptReason RegularPredicateWithUCT::uncommon_trap_reason(IfProj
 }
 
 bool RegularPredicateWithUCT::is_predicate(Node* maybe_success_proj) {
-  if (may_be_predicate_if(maybe_success_proj)) {
-    IfProjNode* success_proj = maybe_success_proj->as_IfProj();
-    const Deoptimization::DeoptReason deopt_reason = uncommon_trap_reason(success_proj);
-    return (deopt_reason == Deoptimization::Reason_loop_limit_check ||
-            deopt_reason == Deoptimization::Reason_predicate ||
-            deopt_reason == Deoptimization::Reason_profile_predicate);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.cpp:98]: extracted to new method `has_valid_uncommon_trap()`

File: src/hotspot/share/opto/predicates.cpp [hunk 8]
```diff
@@ -461,17 +450,45 @@ void InitializedAssertionPredicate::create_halt_node(IfFalseNode* fail_proj, Ide
...
+    PredicatePrinter printer(prefix);
+    PredicateBlockIterator iterator(_tail, _deopt_reason);
+    iterator.for_each(printer);
+  } else {
+    tty->print_cr("%s- <empty>", prefix);
+  }
+}
+
+// Dumps all predicates from the loop to the earliest predicate in a pretty format.
+void Predicates::dump() const {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/predicates.cpp:469]: Example output:
```
239 OuterStripMinedLoop:
- Loop Limit Check Predicate Block:
  - Parse Predicate: 115 ParsePredicate
  - Runtime Predicate: 276 If
- Profiled Loop Predicate Block:
  - Parse Predicate: 104 ParsePredicate
- Loop Predicate Block:
  - Parse Predicate: 93 ParsePredicate
  - Template Assertion Predicate: 270 RangeCheck
  - Template Assertion Predicate: 260 RangeCheck
  - Runtime Predicate: 253 RangeCheck
  - Runtime Predicate: 283 RangeCheck
```

**Review by abdelhak-zaaim — APPROVED** [review]

File: src/hotspot/share/opto/loopnode.cpp [hunk 9]
```diff
@@ -4319,13 +4319,21 @@ void PhaseIdealLoop::mark_loop_associated_parse_predicates_useful() {
   }
 }
 
+// This visitor marks all visited Parse Predicates useful.
+class ParsePredicateUsefulMarker : public PredicateVisitor {
+ public:
+  using PredicateVisitor::visit;
```
Comment (rwestrel) [inline src/hotspot/share/opto/loopnode.cpp:4325]: Why is this needed?

Comment (chhagedorn) [reply src/hotspot/share/opto/loopnode.cpp:4325]: Forgot to explain the reason behind. Without this, compilation fails with:
```
/home/christian/jdk3/open/src/hotspot/share/opto/predicates.hpp:240:16: error: 'virtual void PredicateVisitor::visit(const InitializedAssertionPredicate&)' was hidden [-Werror=overloaded-virtual=]
  240 |   virtual void visit(const InitializedAssertionPredicate& initialized_assertion_predicate) {}
      |                ^~~~~
/home/christian/jdk3/open/src/hotspot/share/opto/loopnode.cpp:4327:8: note:   by 'virtual void ParsePredicateUsefulMarker::visit(const ParsePredicate&)'
 4327 |   void visit(const ParsePredicate& parse_predicate) override {
      |  
```
I was not quite sure what this means, so I looked it up and found this:
https://stackoverflow.com/questions/18515183/c-overloaded-virtual-function-warning-by-clang

It looks like the warning is here to prevent accidental hiding of overloads. Here is another example:
```
struct Base {
  virtual void foo(double e) {
    printf("Base");
  }
};

struct Derived: public Base {
  virtual void foo(int e) {
    printf("Derived");
  }
};

int main() {
  Derived derived;
  derived.foo(3.4f);
  return 0;
}
```
You would expect that this prints `Base`. But since the lookup happens on `Derived`, it finds a matching `Derived::foo(int e)` and applies that one (we print "Derived") even though we intended to call `Base::foo()`. By adding `using Base::foo`, we make the overloaded version available in `Derived` and pick that one instead when running `main()` (we print "Base", as anticipated).

In my example, it would not be a problem. But I guess the warning is here to just generally warn the user if the code was intended or just a typo resulting in unexpected behavior.

**Comment by chhagedorn** [comment]
Thanks for your review!

**Comment by chhagedorn** [comment]
Anyone for a second review?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
Thanks Tobias for your review!

/integrate
